File: sync-service/now_playing.py
"""Now-playing info from the Windows System Media Transport Controls (SMTC).

Uses the winsdk projection of ``Windows.Media.Control`` to read whatever media
session is currently active (Spotify, browsers, groove, etc.) and to send
transport commands. Every public function degrades gracefully: on any failure
(non-Windows host, no winsdk, no active session) it returns
``{"available": False}`` instead of raising, so the gadget can simply hide.
"""

from __future__ import annotations

import asyncio
import base64

_MEDIA_STATUS = {
    0: "closed",
    1: "opened",
    2: "changing",
    3: "stopped",
    4: "playing",
    5: "paused",
}


def _seconds(value) -> float | None:
    try:
        return value.total_seconds()
    except Exception:
        return None


async def _read_thumbnail(props) -> str | None:
    reference = getattr(props, "thumbnail", None)
    if reference is None:
        return None
    try:
        from winsdk.windows.storage.streams import DataReader

        stream = await reference.open_read_async()
        size = stream.size
        if not size or size > 4_000_000:
            return None
        reader = DataReader(stream.get_input_stream_at(0))
        await reader.load_async(size)
        buffer = bytearray(size)
        reader.read_bytes(buffer)
        content_type = stream.content_type or "image/png"
        encoded = base64.b64encode(bytes(buffer)).decode("ascii")
        return f"data:{content_type};base64,{encoded}"
    except Exception:
        return None


async def _get_now_playing_async() -> dict:
    from winsdk.windows.media.control import (
        GlobalSystemMediaTransportControlsSessionManager as MediaManager,
    )

    manager = await MediaManager.request_async()
    session = manager.get_current_session()
    if session is None:
        return {"available": False}

    props = await session.try_get_media_properties_async()
    playback = session.get_playback_info()
    status = int(playback.playback_status)
    timeline = session.get_timeline_properties()

    start = _seconds(timeline.start_time) or 0.0
    end = _seconds(timeline.end_time) or 0.0
    position = (_seconds(timeline.position) or 0.0) - start
    duration = (end - start) if end > start else None

    return {
        "available": True,
        "title": props.title or "",
        "artist": props.artist or "",
        "album": props.album_title or "",
        "status": _MEDIA_STATUS.get(status, "unknown"),
        "playing": status == 4,
        "position": max(0.0, position),
        "duration": duration,
        "appId": session.source_app_user_model_id or "",
        "thumbnail": await _read_thumbnail(props),
    }


async def _control_async(action: str) -> bool:
    from winsdk.windows.media.control import (
        GlobalSystemMediaTransportControlsSessionManager as MediaManager,
    )

    manager = await MediaManager.request_async()
    session = manager.get_current_session()
    if session is None:
        return False

    if action == "play_pause":
        return bool(await session.try_toggle_play_pause_async())
    if action == "next":
        return bool(await session.try_skip_next_async())
    if action == "previous":
        return bool(await session.try_skip_previous_async())
    return False


def get_now_playing() -> dict:
    try:
        return asyncio.run(_get_now_playing_async())
    except Exception:
        return {"available": False}


def media_control(action: str) -> bool:
    try:
        return asyncio.run(_control_async(action))
    except Exception:
        return False

File: sync-service-rs/src/now_playing.rs
use serde_json::{json, Value};

pub fn get_now_playing() -> Value {
    #[cfg(windows)]
    {
        match windows_now_playing() {
            Ok(value) => value,
            Err(_) => json!({ "available": false }),
        }
    }
    #[cfg(not(windows))]
    {
        json!({ "available": false })
    }
}

pub fn media_control(action: &str) -> bool {
    #[cfg(windows)]
    {
        windows_media_control(action).unwrap_or(false)
    }
    #[cfg(not(windows))]
    {
        let _ = action;
        false
    }
}

#[cfg(windows)]
fn windows_now_playing() -> anyhow::Result<Value> {
    use windows::Media::Control::GlobalSystemMediaTransportControlsSessionManager;

    let manager = GlobalSystemMediaTransportControlsSessionManager::RequestAsync()?.get()?;
    let session = manager.GetCurrentSession()?;
    // GetCurrentSession returns null session as error in some bindings — check via Result
    let props = session.TryGetMediaPropertiesAsync()?.get()?;
    let playback = session.GetPlaybackInfo()?;
    let status = playback.PlaybackStatus()?.0;
    let timeline = session.GetTimelineProperties()?;

    let start = timeline.StartTime()?.Duration as f64 / 10_000_000.0;
    let end = timeline.EndTime()?.Duration as f64 / 10_000_000.0;
    let position_raw = timeline.Position()?.Duration as f64 / 10_000_000.0;
    let position = (position_raw - start).max(0.0);
    let duration = if end > start { Some(end - start) } else { None };

    let status_name = match status {
        0 => "closed",
        1 => "opened",
        2 => "changing",
        3 => "stopped",
        4 => "playing",
        5 => "paused",
        _ => "unknown",
    };

    let thumbnail = read_thumbnail(&props).ok().flatten();

    Ok(json!({
        "available": true,
        "title": props.Title().map(|s| s.to_string()).unwrap_or_default(),
        "artist": props.Artist().map(|s| s.to_string()).unwrap_or_default(),
        "album": props.AlbumTitle().map(|s| s.to_string()).unwrap_or_default(),
        "status": status_name,
        "playing": status == 4,
        "position": position,
        "duration": duration,
        "appId": session.SourceAppUserModelId().map(|s| s.to_string()).unwrap_or_default(),
        "thumbnail": thumbnail,
    }))
}

#[cfg(windows)]
fn read_thumbnail(
    props: &windows::Media::Control::GlobalSystemMediaTransportControlsSessionMediaProperties,
) -> anyhow::Result<Option<String>> {
    use base64::Engine;
    use windows::Storage::Streams::DataReader;

    let reference = props.Thumbnail()?;
    let stream = reference.OpenReadAsync()?.get()?;
    let size = stream.Size()?;
    if size == 0 || size > 4_000_000 {
        return Ok(None);
    }
    let reader = DataReader::CreateDataReader(&stream.GetInputStreamAt(0)?)?;
    reader.LoadAsync(size as u32)?.get()?;
    let mut buffer = vec![0u8; size as usize];
    reader.ReadBytes(&mut buffer)?;
    let content_type = stream
        .ContentType()
        .map(|s| s.to_string())
        .unwrap_or_else(|_| "image/png".into());
    let encoded = base64::engine::general_purpose::STANDARD.encode(&buffer);
    Ok(Some(format!("data:{content_type};base64,{encoded}")))
}

#[cfg(windows)]
fn windows_media_control(action: &str) -> anyhow::Result<bool> {
    use windows::Media::Control::GlobalSystemMediaTransportControlsSessionManager;

    let manager = GlobalSystemMediaTransportControlsSessionManager::RequestAsync()?.get()?;
    let session = manager.GetCurrentSession()?;
    let ok = match action {
        "play_pause" => session.TryTogglePlayPauseAsync()?.get()?,
        "next" => session.TrySkipNextAsync()?.get()?,
        "previous" => session.TrySkipPreviousAsync()?.get()?,
        _ => false,
    };
    Ok(ok)
}
